File: java/src/main/java/cn/az/code/other/array/MaximumSubArray.java
package cn.az.code.other.array;

/**
 * @author az
 * @since 10/12/20
 */
public class MaximumSubArray {

    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];//dp[i] means the maximum subarray ending with A[i];
        dp[0] = nums[0];
        int max = dp[0];

        for(int i = 1; i < n; i++){
            dp[i] = nums[i] + (Math.max(dp[i - 1], 0));
            max = Math.max(max, dp[i]);
        }

        return max;
    }
}

File: code/cpp/2021/nov/maximum-subarray.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        size_t sz = nums.size();
        vector<int> v(sz);

        v[0] = nums[0];
        int _max = v[0];
        for (auto i = 1; i < n; ++i)
        {
            v[i] = nums[i] + (v[i - 1] ? 0 v[i - 1] : 0);
            _max = max(max, v[i]);
        }
        return _max;
    }
};